smithy-model/src/main/java/software/amazon/smithy/model/knowledge/TextIndex.java

```diff
@@ -0,0 +1,151 @@
+            visitedNamespaces.add(shape.getId().getNamespace());
+            getTextInstances(shape, textInstanceList, model);
+        });
+
+        for (String namespace : visitedNamespaces) {
+            textInstanceList.add(TextInstance.builder()
+                    .locationType(TextInstance.TextLocation.NAMESPACE)
+                    .text(namespace)
+                    .build());
+        }
+
+        //no point in allowing the list to change
+        textInstanceList = Collections.unmodifiableList(textInstanceList);
+    }
+
+    public static TextIndex of(Model model) {
+        return model.getKnowledge(TextIndex.class, TextIndex::new);
+    }
+
+    public Collection<TextInstance> getTextInstances() {
+        return textInstanceList;
+    }
+
+    private static void getTextInstances(Shape shape,
+                                           Collection<TextInstance> textInstances,
+                                           Model model) {
+        TextInstance.Builder builder = TextInstance.builder()
+                .locationType(TextInstance.TextLocation.SHAPE)
+                .shape(shape);
+        builder.text(shape.getId().getMember().orElseGet(() -> shape.getId().getName()));
+        textInstances.add(builder.build());
+
+        for (Trait trait : shape.getAllTraits().values()) {
+            Shape traitShape = model.expectShape(trait.toShapeId());
+            getTextInstancesForAppliedTrait(trait.toNode(), trait, shape, textInstances,
+                    new ArrayDeque<>(), model, traitShape);
+        }
+    }
+
+    private static void getTextInstancesForAppliedTrait(Node node,
+                                                          Trait trait,
+                                                          Shape parentShape,
+                                                          Collection<TextInstance> textInstances,
+                                                          Deque<TextInstance.PathElement> propertyPath,
+                                                          Model model,
+                                                          Shape currentTraitPropertyShape) {
+        if (trait.toShapeId().equals(ReferencesTrait.ID)) {
+            //Skip ReferenceTrait because it is referring to other shape names already being checked
+        } else if (node.isStringNode()) {
+            textInstances.add(TextInstance.builder()
+                    .locationType(TextInstance.TextLocation.APPLIED_TRAIT)
+                    .shape(parentShape)
+                    .trait(trait)
+                    .traitPropertyPath(propertyPath)
+                    .text(node.expectStringNode().getValue())
+                    .build());
+        } else if (node.isObjectNode()) {
+            ObjectNode objectNode = node.expectObjectNode();
+            objectNode.getMembers().entrySet().forEach(memberEntry -> {
```

Comment:
This can use `getStringMap()` instead of `getMembers()` to remove the repeated `.getKey().getValue()` calls.